nifi-nar-bundles/nifi-slack-bundle/nifi-slack-processors/src/main/java/org/apache/nifi/services/slack/SlackRecordSink.java

```diff
@@ -0,0 +1,144 @@
+            .displayName("Web Service Client Provider")
+            .description("Controller service to provide HTTP client for communicating with Slack API")
+            .required(true)
+            .identifiesControllerService(WebClientServiceProvider.class)
+            .build();
+    private volatile RecordSetWriterFactory writerFactory;
+    private SlackRestService service;
+
+    @Override
+    public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return Collections.unmodifiableList(Arrays.asList(
+                API_URL,
+                ACCESS_TOKEN,
+                CHANNEL_ID,
+                RECORD_WRITER_FACTORY,
+                WEB_SERVICE_CLIENT_PROVIDER
+        ));
+    }
+
+    @OnEnabled
+    public void onEnabled(final ConfigurationContext context) {
+        writerFactory = context.getProperty(RECORD_WRITER_FACTORY).asControllerService(RecordSetWriterFactory.class);
+        final WebClientServiceProvider webClientServiceProvider = context
+                .getProperty(WEB_SERVICE_CLIENT_PROVIDER)
+                .asControllerService(WebClientServiceProvider.class);
+        final String accessToken = context.getProperty(ACCESS_TOKEN).getValue();
+        final String apiUrl = context.getProperty(API_URL).getValue();
+        service = new SlackRestService(webClientServiceProvider, accessToken, apiUrl);
+    }
+
+    @Override
+    public WriteResult sendData(final RecordSet recordSet, final Map<String, String> attributes, final boolean sendZeroResults) throws IOException {
+        WriteResult writeResult;
+        try (final ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+            try (final RecordSetWriter writer = writerFactory.createWriter(getLogger(), recordSet.getSchema(), out, attributes)) {
+                writer.beginRecordSet();
+                Record record = recordSet.next();
+                while (record != null) {
+                    writer.write(record);
+                    writer.flush();
+                    record = recordSet.next();
+                }
+                writeResult = writer.finishRecordSet();
+                writer.flush();
+            } catch (final SchemaNotFoundException e) {
+                final String errorMessage = String.format("RecordSetWriter could not be created because the schema was not found. The schema name for the RecordSet to write is %s",
+                        recordSet.getSchema().getSchemaName());
+                throw new ProcessException(errorMessage, e);
+            }
+
+            try {
+                final String message = out.toString();
+                final String channel = getConfigurationContext().getProperty(CHANNEL_ID).getValue();
+                service.sendMessageToChannel(message, channel);
+            } catch (final SlackRestServiceException e) {
+                getLogger().error("Failed to send message to Slack.", e);
+                throw new ProcessException(e);
```

Comment:
The logger can be removed and the message can be included in the Exception. Also recommend using the `IOException`.
```suggestion
                throw new IOException("Failed to send messages to Slack", e);
```

Reply:
Making the changes